File: Android/session/src/main/java/com/felipearpa/tyche/session/avatar/infrastructure/AvatarRemoteRepository.kt
package com.felipearpa.tyche.session.avatar.infrastructure

import com.felipearpa.network.NetworkExceptionHandler
import com.felipearpa.tyche.session.avatar.domain.AvatarDataSource
import com.felipearpa.tyche.session.avatar.domain.AvatarRepository
import com.felipearpa.tyche.session.avatar.domain.AvatarUploadDataSource
import com.felipearpa.tyche.session.avatar.domain.AvatarUploadUrlRequest

internal class AvatarRemoteRepository(
    private val avatarDataSource: AvatarDataSource,
    private val avatarUploadDataSource: AvatarUploadDataSource,
    private val networkExceptionHandler: NetworkExceptionHandler,
) : AvatarRepository {
    override suspend fun issueUploadUrl(accountId: String, contentLength: Long): Result<String> =
        networkExceptionHandler.handle {
            avatarDataSource.issueUploadUrl(
                accountId = accountId,
                request = AvatarUploadUrlRequest(contentLength = contentLength),
            ).url
        }

    override suspend fun upload(url: String, data: ByteArray): Result<Unit> =
        networkExceptionHandler.handle {
            avatarUploadDataSource.upload(url = url, data = data)
        }
}

File: iOS/Session/Sources/Session/Avatar/Infrastructure/AvatarRemoteRepository.swift
import Core
import Foundation

class AvatarRemoteRepository: AvatarRepository {
    private let avatarRemoteDataSource: AvatarRemoteDataSource
    private let avatarUploadDataSource: AvatarUploadDataSource
    private let networkErrorHandler: NetworkErrorHandler

    init(
        avatarRemoteDataSource: AvatarRemoteDataSource,
        avatarUploadDataSource: AvatarUploadDataSource,
        networkErrorHandler: NetworkErrorHandler
    ) {
        self.avatarRemoteDataSource = avatarRemoteDataSource
        self.avatarUploadDataSource = avatarUploadDataSource
        self.networkErrorHandler = networkErrorHandler
    }

    func issueUploadUrl(accountId: String, contentLength: Int) async -> Result<URL, Error> {
        return await networkErrorHandler.handle {
            let response = try await avatarRemoteDataSource.issueUploadUrl(
                accountId: accountId,
                request: AvatarUploadUrlRequest(contentLength: contentLength)
            )

            guard let url = URL(string: response.url) else {
                throw NetworkError.remoteCommunication
            }

            return url
        }
    }

    func upload(_ data: Data, to url: URL) async -> Result<Void, Error> {
        return await networkErrorHandler.handle {
            try await avatarUploadDataSource.upload(data, to: url)
        }
    }
}
